feat: Add POST event functions based for v2 API

lindseysimple opened the pull request:
Add POST event controller/handler/db functions based on v2 API.
Fixes: #2609

Signed-off-by: Lindsey Cheng <beckysocute@gmail.com>

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit subject follows the [Conventional Commits spec](https://github.com/zeke/semantic-pull-requests)
- [x] The commit message follows the [EdgeX Contributor Guide](https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] `make test` has completed successfully

## PR Type

What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
<!-- Please describe the current behavior and link to a relevant issue. -->

Issue Number: #2609

## What is the new behavior?
The v2 edgex-go functions are designed under the following architecture:

- Transport (UI) Layer: The controller functions. Transport-specific logic implementation, such as REST(HTTP) or MQTT.
- Handler (Application) Layer: The handler functions. Business logic or use-case implementation.
- Infrastructure (Persistence) Layer: The infrastructure functions. Repository contract and implementation for different persistences.

New **controller**/**handler**/**infrastructure** functions are added under `./internal/pkg/v2` path for POST Event based on the v2 OpenAPI swagger [file](https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-data/2.x#/default/post_event) definitions and v2 core data DTO/model definitions (added in this `go-mod-core-contracts` [PR](https://github.com/edgexfoundry/go-mod-core-contracts/pull/236)).

## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [ ] No

## Are there any new imports or modules? If so, what are they used for and why?
No

## Are there any specific instructions or things that should be known prior to reviewing?

## Other information

lenny-intel commented on internal/pkg/v2/bootstrap/container/config.go:
```diff
@@ -0,0 +1,20 @@
+package container
+
+import (
+	"github.com/edgexfoundry/edgex-go/internal/core/data/config"
```
This is core data specific, but in the edgex-go common code. This needs to be changed to not uses core data's specific ConfigurationStruct. There is an interface in go-mod-bootstrap/config/configuration designed to handle the config structs generically.

lindseysimple replied:
Move to `/internal/core/data/v2 `.

lenny-intel commented on internal/pkg/v2/bootstrap/container/container.go:
```diff
@@ -0,0 +1,56 @@
+)
+
+type Container struct {
+	ConfigClient       *config.ConfigurationStruct
```
Same issue with using core data's specific ConfigurationStruct.

lindseysimple replied:
Move to `/internal/core/data/v2 `.

lenny-intel commented on internal/pkg/v2/bootstrap/container/container.go:
```diff
@@ -0,0 +1,56 @@
+	MetadataClient     metadata.DeviceClient
+}
+
+var BootstrapContainer *di.Container
```
Please do not use global variables. Use structs and constructor and receiver functions to avoid global variables.
example:
```
type MyThing struct {
  BootstrapContainer *di.Container
  RegistryClients Container
}

func NewMyThing() MyThing {
   return MyThing{}
}

func (thing MyThing) Init() {
  ...
}
```

lindseysimple replied:
Hi Lenny, the above code has been refactor in [internal/core/data/v2/bootstrap/container/serviceinjector.go](https://github.com/edgexfoundry/edgex-go/blob/9bb6161d8711a2382a3e6df77ea55c1240982a91/internal/core/data/v2/bootstrap/container/serviceinjector.go),  could you please review the change? Thanks.

lenny-intel commented on internal/pkg/v2/bootstrap/container/container.go:
```diff
@@ -0,0 +1,56 @@
+)
+
+type Container struct {
+	ConfigClient       *config.ConfigurationStruct
```
This is not a client, so not found of the name `ConfigClient`

lindseysimple replied:
Updated.

lenny-intel commented on internal/pkg/v2/bootstrap/container/container.go:
```diff
@@ -0,0 +1,56 @@
+type Container struct {
+	ConfigClient       *config.ConfigurationStruct
+	DBClient           interfaces.DBClient
+	ErrorHandlerClient *errorconcept.Handler
```
Remove `Client` from the name.

lenny-intel commented on internal/pkg/v2/bootstrap/container/container.go:
```diff
@@ -0,0 +1,56 @@
+	"github.com/edgexfoundry/go-mod-messaging/messaging"
+)
+
+type Container struct {
```
Not wild about the generic name `Container`. What is this `thing` represent? File name should also be more specific.

lindseysimple replied:
Rename from `Container` to `ServiceInjector` as this struct is used to save all the service clients injected by `go-mod-bootstrap`. Also rename the file to `serviceinjector.go`.

lenny-intel commented on internal/pkg/v2/bootstrap/container/container.go:
```diff
@@ -0,0 +1,56 @@
+var RegistryClients Container
+
+func Init() {
+	dic := BootstrapContainer
```
Where is` BootstrapContainer` initialized? If uses receiver approach it can be done in the New func. Also, Why the need to reassign it to `dic`

lindseysimple replied:
Refactor this part with a new v2 `BootstrapHandler` separated from v1, so the above code has been removed.

lenny-intel commented on internal/pkg/v2/bootstrap/container/container.go:
```diff
@@ -0,0 +1,56 @@
+
+func Init() {
+	dic := BootstrapContainer
+	chEvents := make(chan interface{}, 100)
```
Create constant for `100`.

lindseysimple replied:
Removed the `chEvents` as it is no longer needed in v2 API to receive the `DeviceLastReported` and `DeviceServiceLastReported` events.

lenny-intel commented on internal/pkg/v2/controller/errorconcept/default.go:
```diff
@@ -0,0 +1,110 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package errorconcept
```
Why name `errorconcept`? what does `concept` mean in this domain? These are custom errors, so why not just `error` or `errors`

lindseysimple replied:
The `errorconcept` package is inherited from v1 [internal/pkg/errorconcept/handler.go](https://github.com/edgexfoundry/edgex-go/blob/master/internal/pkg/errorconcept/handler.go), and we'll be happy to rename the package name if needed.

lenny-intel commented on internal/pkg/v2/bootstrap/handlers/database/database.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+// NewDatabaseForCoreData is a factory method that returns an initialized Database receiver struct.
+func NewDatabaseForCoreData(httpServer httpServer, database interfaces.Database) Database {
```
Again we have Core Data specific code in the common package. This should be in edgex-go/internal/core/data/v2/...

lindseysimple replied:
Moved to `internal/core/data/v2/`.

lindseysimple commented on cmd/core-command/Attribution.txt:
```diff
@@ -162,3 +162,15 @@ https://github.com/cloudflare/gokey/blob/master/LICENSE
+https://github.com/go-playground/locales/blob/master/LICENSE
+
+go-playground/universal-translator (MIT) https://github.com/go-playground/universal-translator
+https://github.com/go-playground/universal-translator/blob/master/LICENSE
```
Yes, as the `playground/universal-translator` and `go-playground/locales` packages are used in `go-playground/validator`.

lindseysimple commented on internal/core/data/v2/controller/http/core/data/event.go:
```diff
@@ -0,0 +1,76 @@
+	respDTO "github.com/edgexfoundry/go-mod-core-contracts/v2/dtos/responses"
+)
+
+func EventController(w http.ResponseWriter, r *http.Request, dic *di.Container) {
```
@lenny-intel Lenny, as we proposed the v2 API design architecture, we separate the code structure into 3 layers:

- Transport (UI) Layer: The controller functions. Transport-specific logic implementation, such as REST(HTTP) or MQTT.
- Handler (Application) Layer: The handler functions. Business logic or use-case implementation.
- Infrastructure (Persistence) Layer: The infrastructure functions. Repository contract and implementation for different persistences.

The `EventController` function handles the HTTP request, processes the content as input to the controller function, therefore the naming here is `Controller`.
If any other suggestion or concern about this, please let us know, thanks.

lenny-intel replied:
My point is the  standard `Handler` naming convention, as you have used in the description above. `Controller` to me and is suspect others has meaning that has more responsibility that a simple `Handler`. When we see the name with `Handler` in it we know right away it is for a specific REST endpoint.

cloudxxx8 replied:
Currently, in `Handler` signature, there is no protocol specific argument, such as httpRequest or httpResponse.
It will be a general business logic layer for the future service interaction protocols (different Controllers).
If we rename `Handler` to `Processor`, is it ok?

lindseysimple commented on internal/core/data/v2/bootstrap/container/config.go:
```diff
@@ -0,0 +1,20 @@
+//
```
Removed this file and use the existing one ` internal/core/data/container/config.go.`

lindseysimple commented on internal/core/data/v2/bootstrap/container/errorconcept.go:
```diff
@@ -0,0 +1,20 @@
+package container
+
+import (
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/errorconcept"
```
Rename the `errorconcept` package to `error` in v2.

lindseysimple commented on internal/core/data/v2/bootstrap/container/events.go:
```diff
@@ -0,0 +1,21 @@
+//
```
Removed this file as the event channel is no longer needed in v2.

lindseysimple commented on internal/core/data/v2/bootstrap/container/messaging.go:
```diff
@@ -0,0 +1,20 @@
+//
```
Removed this file and use the existing one `internal/core/data/container/messaging.go`.

lindseysimple commented on internal/core/data/v2/bootstrap/container/metadata.go:
```diff
@@ -0,0 +1,20 @@
+//
```
@lenny-intel Lenny, I've discussed this with @cloudxxx8 and he suggests that we still keep a separate `metadata.DeviceClient` for v2 as the design of `metadata.DeviceClient` will be changing in v2 as well.
Please let us know if this makes sense and any other concern or suggestion you have, thanks.

lenny-intel replied:
Right... So you will need to give it a V2 specific name then so the v1 and v2 can coexist in the `dic`

lindseysimple commented on internal/core/data/v2/controller/http/core/data/event.go:
```diff
@@ -0,0 +1,76 @@
+	}
+
+	// retrieve all the service injections from bootstrap
+	httpErrorHandler := v2ServiceInjector.ErrorHandler
```
Updated with `dic`, also removed the second `core/data`  in the file path due to redundancy.

lindseysimple commented on internal/core/data/v2/bootstrap/container/serviceinjector.go:
```diff
@@ -0,0 +1,42 @@
+//
```
Removed `serviceinjector.go` as suggested.

lindseysimple commented on internal/core/data/v2/handler/common/encoding.go:
```diff
@@ -0,0 +1,27 @@
+//
```
Use the existing one in `internal/pkg/encoding.go` instead.

lindseysimple commented on internal/core/data/v2/handler/common/io.go:
```diff
@@ -0,0 +1,45 @@
+
+// EventReader unmarshals a request body into an Event type
+type EventReader interface {
+	//Read(reader io.Reader, ctx *context.Context) ([]model.Event, error)
```
Removed.

lindseysimple commented on internal/core/data/v2/handler/common/io.go:
```diff
@@ -0,0 +1,45 @@
+// EventReader unmarshals a request body into an Event type
+type EventReader interface {
+	//Read(reader io.Reader, ctx *context.Context) ([]model.Event, error)
+	Read(reader io.Reader, ctx *context.Context) ([]dto.AddEventRequest, error)
```
Renamed to `ReadAddEventRequest`.

lindseysimple commented on internal/core/data/v2/handler/common/io.go:
```diff
@@ -0,0 +1,45 @@
+	return addEvents, nil
+}
+
+// NewJsonReader creates a new instance of cborReader.
```
Re-order as suggested.

lindseysimple commented on internal/core/data/v2/handler/common/io.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+// NewJsonReader creates a new instance of cborReader.
+func NewJsonReader() jsonReader {
```
Renamed.

lindseysimple commented on internal/core/data/v2/handler/common/io.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+// NewRequestReader returns a BodyReader capable of processing the request body
+func NewRequestReader() EventReader {
```
Renamed.

lindseysimple commented on internal/core/data/v2/handler/core/data/device.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+// This function will be updated when CheckDevice in v2 core-metadata is available
+func checkDevice(device string, ctx context.Context, si *v2container.ServiceInjector) error {
```
Use `dic` instead of `si` as suggested.

lindseysimple commented on internal/core/data/v2/controller/http/core/data/event.go:
```diff
@@ -0,0 +1,76 @@
+package data
```
Removed the second `core/data` in the file path due to redundancy.

lindseysimple commented on internal/core/data/v2/handler/common/io.go:
```diff
@@ -0,0 +1,45 @@
+//
```
Updated the file name as suggested, thanks.

lindseysimple commented on internal/core/data/v2/handler/core/data/event.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+// Put event on the message queue to be processed by the rules engine
+func putEventOnQueue(evt model.Event, ctx context.Context, si *v2container.ServiceInjector) {
```
Updated.

lindseysimple commented on internal/core/data/v2/bootstrap/init.go:
```diff
@@ -0,0 +1,113 @@
+//
```
Removed this file as suggested.

cloudxxx8 commented:
@lindseysimple we renamed the device field in event and reading to deviceName.  please upgrade to the latest contract when you are available.

lindseysimple commented on api/openapi/v2/core-data.yaml:
```diff
@@ -16,7 +16,7 @@ components:
       type: object
       properties:
-        device:
+        deviceName:
```
Updated

lindseysimple commented on api/openapi/v2/core-data.yaml:
```diff
@@ -990,6 +988,37 @@ paths:
             application/json:
               schema:
                 $ref: '#/components/schemas/ErrorResponse'
+  /reading/count/device/{deviceId}:
```
OK, got it.

lindseysimple commented on internal/core/data/init.go:
```diff
@@ -124,6 +126,12 @@ func (b *Bootstrap) BootstrapHandler(ctx context.Context, wg *sync.WaitGroup, _
 		errorContainer.ErrorHandlerName: func(get di.Get) interface{} {
 			return errorconcept.NewErrorHandler(lc)
 		},
+		v2DataContainer.MetadataDeviceClientName: func(get di.Get) interface{} { // add v2 API MetadataDeviceClient
+			return mdc
+		},
+		v2DataContainer.ErrorHandlerName: func(get di.Get) interface{} { // add v2 API error handler
+			return error.NewErrorHandler(lc)
+		},
```
Updated v2DataContainer.MetadataDeviceClientName to `V2MetadataDeviceClient` to avoid the name conflict.

lindseysimple commented on internal/core/data/router.go:
```diff
@@ -477,6 +478,15 @@ func loadRestRoutes(r *mux.Router, dic *di.Container) {
 
+	// v2 API routes
+	// Events
+	r.HandleFunc(v2Constant.ApiEventRoute, func(w http.ResponseWriter, r *http.Request) {
```
Separate the v2 API routes in `v2/router.go` as suggested.

lindseysimple commented on internal/core/data/v2/infrastructure/client.go:
```diff
@@ -0,0 +1,110 @@
+//
```
Updated the path to `.../infrastructure /redis/client.go`

lindseysimple commented on internal/core/data/v2/infrastructure/db.go:
```diff
@@ -0,0 +1,37 @@
+//
```
Removed this file as suggested.

lindseysimple commented on internal/core/data/v2/infrastructure/client.go:
```diff
@@ -0,0 +1,110 @@
+				"tcp", connectionString, opts...,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("Could not dial Redis: %s", err)
+			}
+			return conn, nil
+		}
+		// Default the batch size to 1,000 if not set
+		batchSize := 1000
+		if config.BatchSize != 0 {
+			batchSize = config.BatchSize
+		}
+		currClient = &CoreDataClient{
+			Pool: &redis.Pool{
+				IdleTimeout: 0,
+				/* The current implementation processes nested structs using concurrent connections.
+				 * With the deepest nesting level being 3, three shall be the number of maximum open
+				 * idle connections in the pool, to allow reuse.
+				 * TODO: Once we have a concurrent benchmark, this should be revisited.
+				 * TODO: Longer term, once the objects are clean of external dependencies, the use
+				 * of another serializer should make this moot.
+				 */
+				MaxIdle: 10,
+				Dial:    dialFunc,
+			},
+			BatchSize: batchSize,
+			logger:    lc,
+		}
+	})
+	return currClient, nil
+}
+
+// CloseSession closes the connections to Redis
+func (c *CoreDataClient) CloseSession() {
+	c.Pool.Close()
+	close(deleteEventsChannel)
+	close(deleteReadingsChannel)
+	currClient = nil
+	once = sync.Once{}
+}
```
Updated using the existing `NewClient` function in `internal/pkg/db/redis/client.go` as suggested.

lindseysimple commented on internal/core/data/v2/controller/http/event.go:
```diff
@@ -0,0 +1,76 @@
+	ctx := r.Context()
+
+	reader := io.NewEventRequestReader()
+	addEventReqDTOs, err := reader.ReadAddEventRequest(r.Body, &ctx)
```
@lenny-intel , 
Hi Lenny, as in swagger document the request body of the POST event is an array of `AddEventRequest` DTOs,
```
  /event:
    post:
      summary: "Allows for the ingestion of event/reading data"
      requestBody:
        required: true
        content:
          application/json:
            schema:
              type: array
              items:
                $ref: '#/components/schemas/AddEventRequest'
```
Therefore the `reader.ReadAddEventRequest` function is designed to read the request body and unmarshal to an addEventRequest DTO array. If there's some misunderstanding here, please let me know, thanks.

lenny-intel replied:
Ahh, ok i missed the `type: array`. works for me. THX

lindseysimple commented on internal/core/data/v2/handler/event.go:
```diff
@@ -0,0 +1,82 @@
+
+	// Add the event and readings to the database
+	if configuration.Writable.PersistData {
+		addedEvent, err := dbClient.AddEvent(e)
```
@lenny-intel  Hi Lenny, as discussed @cloudxxx8 , `dbClient.AddEvent(e)` now returns the event model instead of simply id in v2 API implementation so that we can retrieve the `id`, `created` field added in the db layer as Mike suggested in https://github.com/edgexfoundry/edgex-go/pull/2605#discussion_r451734825. 
It seems that is making more sense that the `created` time stamp is assigned in the db layer (the time when the entry is created in DB).
Would you please let us know what you think about this? Thanks.

lindseysimple commented on internal/pkg/v2/common/util.go:
```diff
@@ -0,0 +1,12 @@
+//
```
@lenny-intel I've updated the file path to `internal/pkg/common/util.go`, or should we remove the common package in the file path and use `internal/pkg/util.go` instead? Thanks.

lenny-intel replied:
`internal/pkg` is common, so probably bets to go with `internal/pkg/util.go`

lindseysimple commented on internal/pkg/v2/error/default.go:
```diff
@@ -0,0 +1,110 @@
+var Default defaultErrorConcept
+
+// DefaultErrorConcept represents a fallback error concept only
+type defaultErrorConcept struct {
```
Updated.

lindseysimple commented on internal/pkg/v2/error/handler.go:
```diff
@@ -0,0 +1,77 @@
+//
```
Updated.

lindseysimple commented on internal/pkg/v2/error/handler.go:
```diff
@@ -0,0 +1,77 @@
+	logger logger.LoggingClient
+}
+
+type SingleHTTPResponse struct {
```
Renamed to `SingleHTTPErrorResponse` as suggested.

lindseysimple replied:
Renamed to `HTTPErrorResponse`.

lindseysimple commented on internal/pkg/v2/error/handler.go:
```diff
@@ -0,0 +1,77 @@
+// HandleOneVariant applies general error-handling with a single allowable error and a default error to be used as a
+// fallback when none of the allowable errors are matched
+func (e *Handler) HandleOneVariant(err error, allowableError ErrorConceptType, defaultError ErrorConceptType) SingleHTTPResponse {
+	message := allowableError.message(err)
+	e.logger.Error(message)
+	return SingleHTTPResponse{
+		Status:  allowableError.httpErrorCode(),
+		Message: message,
+	}
+
```
Updated the `ErrorHandler` interface to have 3 methods `Handle`, `HandleWithDefault` and `HandleManyWithDefault` which is similar to the original `ErrorHandler` interface code structure in `internal/pkg/errorconcept/handler.go`.

lindseysimple commented on internal/pkg/v2/error/handler.go:
```diff
@@ -0,0 +1,77 @@
+// HandleOneVariant applies general error-handling with a single allowable error and a default error to be used as a
+// fallback when none of the allowable errors are matched, call HttpDirectHandle to sent a direct HTTP response
+func (e *Handler) HttpHandleOneVariant(w http.ResponseWriter, err error, allowableError ErrorConceptType, defaultError ErrorConceptType) {
+	if allowableError != nil && allowableError.isA(err) {
```
Function name updated.

lindseysimple commented on internal/pkg/v2/error/service_client.go:
```diff
@@ -0,0 +1,58 @@
+//
```
As we declared a new `ErrorType` interface in `internal/pkg/v2/handler.go`, cannot used the existing `internal/pkg/errorconcept/service_client.go` file as they are implemented the `ErrorConceptType` interface in `internal/pkg/errorconcept/handler.go`.

lenny-intel commented on internal/core/data/v2/controller/http/event.go:
```diff
@@ -0,0 +1,76 @@
+	respDTO "github.com/edgexfoundry/go-mod-core-contracts/v2/dtos/responses"
+)
+
+func AddEventController(w http.ResponseWriter, r *http.Request, dic *di.Container) {
```
If you really want to use `Controller` then you should create a proper event controller object. i.e.

```
type EventController struct {
   lc LoggingClient,
   errorHandler EroorHandler,
   reader EventRequestReader
}


NewEventController(dic *di.Container) *EventController {
   ec := EventController {}
   ec.lc := ...
   ec.errorHandler := 
   ec.reader := 
  return ec
}

func (ec *EventController) AddEvent(w http.ResponseWriter, r *http.Request) {
...
}
```

Then in v2.LoadRestRoutes you call `ec := NewEventController` and use `ec.AddEvent` in call to `HandleFunc`

This is what I expect to see in the code when the term `Controller` is used and I like the code structured this way.

cloudxxx8 replied:
Thanks for the suggestion, @lenny-intel we are adopting this implementation.
I don't insist to use the naming here.  However, we had better decide the proper naming convention at this state.
We can rename `Controller` to `Handler`, but I can't find a good name for the business logic layer (origin `Handler` package).  Do you have any suggestion?
I still think we need three layers.

lenny-intel replied:
We this change I am happy with `Controller`. It is now conceptually appropriately name. THX!!

lindseysimple commented on internal/core/data/v2/bootstrap/container/database.go:
```diff
@@ -0,0 +1,20 @@
+package container
+
+import (
+	"github.com/edgexfoundry/edgex-go/internal/core/data/v2/infrastructure/redis/interfaces"
```
Updated `internal/core/data/v2/infrastructure/redis/interfaces/db.go` to `internal/core/data/v2/infrastructure/interfaces/db.go` as suggested

lindseysimple commented on internal/core/data/v2/infrastructure/redis/interfaces/db.go:
```diff
@@ -0,0 +1,16 @@
+//
```
Removed `redis` from file path as suggested.

lindseysimple commented on internal/core/data/v2/infrastructure/redis/interfaces/mocks/DBClient.go:
```diff
@@ -0,0 +1,40 @@
+// Code generated by mockery v1.0.0. DO NOT EDIT.
```
Removed `redis` from file path as suggested.

lindseysimple commented on internal/core/data/v2/bootstrap/handlers/database/database.go:
```diff
@@ -0,0 +1,126 @@
+		},
+	})
+
+	lc.Info("Database connected")
```
Updated.

lindseysimple commented on internal/core/data/v2/handler/event.go:
```diff
@@ -0,0 +1,82 @@
+	msgEnvelope := msgTypes.NewMessageEnvelope(data, ctx)
+	err = msgClient.Publish(msgEnvelope, configuration.MessageQueue.Topic)
+	if err != nil {
+		lc.Error(fmt.Sprintf("Unable to send message for event: %+v %v", evt, err))
```
Updated as suggested.

lindseysimple commented on internal/pkg/v2/error/handler.go:
```diff
@@ -0,0 +1,67 @@
+// HandleWithDefault applies general error-handling with a single allowable error and a default error to be used as a
+// fallback when none of the allowable errors are matched
+func (e *Handler) HandleWithDefault(err error, allowableError ErrorType, defaultError ErrorType) HTTPErrorResponse {
+	if allowableError != nil && allowableError.isA(err) {
```
Updated the `isA` function in `internal/pkg/v2/error/service_client.go` to 
```
func (r errContractInvalid) isA(err error) bool {
	_, ok := err.(v2.ErrContractInvalid)
	return ok
}
```
to match the conversion type of the receiver.

lindseysimple commented on internal/core/data/v2/controller/http/event.go:
```diff
@@ -0,0 +1,75 @@
+package http
+
+import (
+	"github.com/edgexfoundry/edgex-go/internal/core/data/v2/handler"
```
Updated.

lenny-intel commented on internal/core/data/v2/bootstrap/handlers/database/database.go:
```diff
@@ -0,0 +1,131 @@
+type Database struct {
+	httpServer httpServer
+	database   interfaces.Database
+	isCoreData bool
```
Why do we need this?

lindseysimple replied:
In v1 implementation, the `isCoreData` boolean is used to create a new CoreData db client in https://github.com/edgexfoundry/edgex-go/blob/c5134682d316d7e5a1de6694786c0c16bb482985/internal/pkg/db/redis/client.go#L45-L57
We might still need this for V2 API implementation.

lenny-intel replied:
Please remove and add back when actually needed in V2.

lindseysimple replied:
Removed the `isCoreData` bool in `Database` struct of `internal/core/data/v2/bootstrap/handlers/database/database.go`, and replaced `CoreDataClient` struct with a general `Client` in `internal/core/data/v2/infrastucture/redis/client.go`.

lenny-intel commented on internal/core/data/v2/bootstrap/handlers/database/database.go:
```diff
@@ -0,0 +1,131 @@
+			}
+			time.Sleep(time.Second)
+		}
+		lc.Info("Database disconnected")
```
Remove duplicate message

lindseysimple replied:
Updated the message to `Database for V2 API disconnected`.

lenny-intel replied:
Ahh, it is `disconnected` message... ;-) THX!

lenny-intel commented:
@lindseysimple , Great progress BTW!! :-)

lindseysimple commented:
> Still have open questions around:
> 
> 3. Need to understand how the App Services and others receiving Event from Message Bus will determine is it is a V1 or V2 event.

@lenny-intel Lenny, There would be two ways to identify the event is from V1 or V2. One is to add a "V2" suffix to the message queue topic name `configuration.MessageQueue.Topic`.
https://github.com/edgexfoundry/edgex-go/blob/037dd24482fcb1b9af402d44251ed3773dc665ee/internal/core/data/v2/handler/event.go#L72
Another alternative is to check if the `Event` model has the `DeviceName` field which is new in V2 instead of `Device`. Which one do you think is better? Thanks.

cloudxxx8 commented on internal/core/data/v2/infrastructure/event.go:
```diff
@@ -0,0 +1,65 @@
+//
```
I prefer to separate them, but not insist

lenny-intel replied:
It was agreed in the Core WG today to have the DB implementations all in one package under internal/pkg/

lenny-intel commented:
> One is to add a "V2" suffix to the message queue topic name `configuration.MessageQueue.Topic`.

App Services config currently only allows a single topic to subscribe...

> Another alternative is to check if the Event model has the DeviceName field which is new in V2 instead of Device.

IMO that is too arbitrary. Really need something more explicit like `Version`. Doesn't exist in JSON means it is V1 and set to `V2` for V2 JSON. We will have to inspect the JSON before unmarshalling.

lenny-intel commented on internal/core/data/v2/controller/http/event.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+// NewEventController creates and initializes an EventController
+func NewEventController(dic *di.Container) *EventController {
```
Excellent!

cloudxxx8 commented:
> > One is to add a "V2" suffix to the message queue topic name `configuration.MessageQueue.Topic`.
> 
> App Services config currently only allows a single topic to subscribe...

Users can set up multiple App Service for different topics.

lenny-intel commented on internal/core/data/v2/handler/event.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	msgEnvelope := msgTypes.NewMessageEnvelope(data, ctx)
+	err = msgClient.Publish(msgEnvelope, configuration.MessageQueue.Topic)
```
It was agreed in the Core WG that all DTOs will have a version property and that V2 will publish the Event DTO to the message bus rather than the Event model. 

This impacts when the DTO is converted to a model, which should be just before storing it so the DTO is still available to publish to the message bus.

lenny-intel commented on internal/core/data/v2/handler/io/event.go:
```diff
@@ -0,0 +1,44 @@
+//
```
Can we more `io` package up one level so it isn't in `handler`?

lindseysimple replied:
Updated the `io` package to one level up as suggested.

lenny-intel commented on internal/core/data/v2/infrastructure/redis/client.go:
```diff
@@ -0,0 +1,59 @@
+//
```
It was decided in the Core WG meeting that all the redis implementation files should be in same location under internal/pkg/v2/infrastructure/redis similar to how it is V1.

lindseysimple replied:
Updated the path to `internal/pkg/v2/infrastructure/redis`.

lenny-intel commented:
> Users can set up multiple App Service for different topics.

Per Core Wg meeting we agreed the V2 DTOs should all have a version property and the Event DTO should be published to the message bus.

lindseysimple commented on internal/core/data/v2/bootstrap/handlers/database.go:
```diff
@@ -0,0 +1,130 @@
+type Database struct {
+	httpServer httpServer
+	database   interfaces.Database
+	isCoreData bool
```
Removed `isCoreData` boolean in v2 as we no longer have a specific Core Data db client in v2 so removing this for now.

lindseysimple commented on internal/core/data/v2/bootstrap/handlers/database.go:
```diff
@@ -0,0 +1,130 @@
+//
```
Updated the path to `internal/pkg/v2`.

lindseysimple commented on internal/core/data/v2/handler/event.go:
```diff
@@ -0,0 +1,90 @@
+//
```
Updated `handle` package/layer name to use `application` instead as suggested.

lenny-intel commented on internal/pkg/v2/infrastructure/interfaces/db.go:
```diff
@@ -0,0 +1,16 @@
+//
```
As with V1 (https://github.com/edgexfoundry/edgex-go/blob/master/internal/core/data/interfaces/db.go) this is the specific interface for Core Data to use, so it does belong in internal/core/data/...

lindseysimple replied:
Thanks for pointing out and that makes sense! Updated the path to `internal/core/data/v2/infrastructure/interfaces/db.go`.

lenny-intel commented on internal/pkg/v2/infrastructure/redis/event.go:
```diff
@@ -0,0 +1,68 @@
+//
```
@andresrinivasan , needs to review this code.

andresrinivasan replied:
I'm curious why the move from storing the object as a blob to storing the object broken down in a hash. The current query interface won't be able to take advantage of this and when an object is returned, it's the entire object, not a subset of fields.

lindseysimple replied:
The reason that we used hashes is that it looks like a good way to represent a data model(a Golang Struct data type) as a Redis data structure. The `AddFlat` and `ScanStruct` methods in `github.com/gomodule/redigo/redis` library is easy to deal with structs for the `HSET` and `HGET` redis commands so event if we don't need to get a specific key in the object, we can still easily to convert between Hashes and models. 
Please let us if any other suggestion or concern, thanks.

andresrinivasan replied:
Agreed that the Hash data structure can be used to mirror the Go struct. From the perspective of the EdgeX API, we mostly query on indexes that aren't well represented by a Hash and we always return the entire struct. That's why the current implementation denormalizes the data into multiple indexes that all point to the same blob that will be returned.

lenny-intel replied:
@lindseysimple , @cloudxxx8 , I discussed the merits between using `Hash` and using `Blob` with @andresrinivasan today.

The use of `hash`, while functional,  is not required for our implementation and is slightly less efficient. One benefit of staying with use of `Blob`, as done for V1, is we can easily review the new V2 code using the V1 code as the standard, without needing @andresrinivasan to participate, even when there are slight differences in the V2 implementation's use of `Blob`.

Thoughts?

lindseysimple replied:
@lenny-intel Lenny, Thanks for the update. Discussed with @cloudxxx8 today and we agree that save as `Blob` will be more suitable for the case here since we return the entire object in the API response instead of a single field. 

Already updated the infrastructure layer functions to save as blob as in v1.

lenny-intel commented on internal/pkg/v2/infrastructure/redis/reading.go:
```diff
@@ -0,0 +1,63 @@
+//
```
@andresrinivasan needs to review this code.

andresrinivasan replied:
Apologies @lenny-intel. Will get to this later this week.

andresrinivasan replied:
Ditto on the curiosity point from `redis.event`.

The duplication of code in the two case sections caught my attention. I suggest a refactor.

The error for the `default:` case is currently using `ErrUnsupportedDatabase`. I don't believe this error is an unknown database type error.

lindseysimple replied:
As model.Reading needs to be converted to two concrete Reading structs `model.BinaryReading` and `model.SimpleReading` and then the value can be retrieved from the structs, it's a bit difficult here to not avoid duplicate code. Would you please suggest if any work around?

As for the `default` case, a new error type `db.ErrUnsupportedReading` is added for the case. Thanks a lot for catching this issue.

andresrinivasan replied:
I suggest refactoring the common code in to a function.

lenny-intel replied:
I agree, majority of the code is identical in each case, just create the `newReading` in each case and then move the common code to after the switch, before the return.

cloudxxx8 replied:
ok, we are refactoring it

lenny-intel commented on internal/core/data/v2/application/event.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// Put event on the message queue to be processed by the rules engine
+func putEventOnQueue(evt model.Event, ctx context.Context, dic *di.Container) {
```
This needs to publish the Event DTO, not Event model, to the message bus. The DTO will have the API version from https://github.com/edgexfoundry/go-mod-core-contracts/pull/255

lindseysimple replied:
Will update this until https://github.com/edgexfoundry/go-mod-core-contracts/pull/255 is merged.

lindseysimple replied:
Updated with Event DTO to publish on the message bus.

lindseysimple commented:
Removed the swagger file `api/openapi/v2/core-data.yaml` from this PR as the changes are now addressed in  PR https://github.com/edgexfoundry/edgex-go/pull/2631.

lenny-intel commented on internal/core/data/v2/application/event.go:
```diff
@@ -0,0 +1,93 @@
+	msgClient := dataContainer.MessagingClientFrom(dic.Get)
+	configuration := dataContainer.ConfigurationFrom(dic.Get)
+
+	lc.Info("Putting event on message queue")
```
Add V2 to the message so we know in the logs when running with both V1 and V2

lindseysimple replied:
Updated.

lenny-intel commented on internal/core/data/v2/application/event.go:
```diff
@@ -0,0 +1,93 @@
+	msgEnvelope := msgTypes.NewMessageEnvelope(data, ctx)
+	err = msgClient.Publish(msgEnvelope, configuration.MessageQueue.Topic)
+	if err != nil {
+		lc.Error(fmt.Sprintf("Unable to send message for event. Correlation-id: %s, Device Name: %s, Error: %v", correlation.FromContext(ctx), evt.DeviceName, err))
```
Add V2 to the message so we know in the logs when running with both V1 and V2

lindseysimple replied:
Updated.

lindseysimple commented on internal/core/data/v2/application/device_test.go:
```diff
@@ -0,0 +1,69 @@
+
+// Supporting methods
+// Reset() re-initializes dependencies for each test
+func reset() {
```
> Expected to see tests related to device, but looks more like test helper functions for the package. Rename the file more appropriate?

Renamed to `test_utils.go`.

cloudxxx8 replied:
we need xxx_test.go for the helper functions, and it won't be included to compiled binary.  These helper functions are for event only.  Why do we move to another file?

cloudxxx8 replied:
@lindseysimple we can test checkDevice() by passing different `configuration.Writable.MetaDataCheck` to see whether it really call `mdc.CheckForDevice`

lenny-intel replied:
@cloudxxx8 , good point on the proper naming of the file. My issue with the original name was that the contents it didn't have anything to do with testing `device`. Actual no unit tests at all.

lenny-intel replied:
Just got @cloudxxx8 point about these for event tests. Should these simply put into the `event_test.go` file and remove this file rather than renaming again?

lindseysimple replied:
Updated to put those util functions in `event_test.go` and removed `test_utils.go`. Thanks.

lindseysimple commented on internal/core/data/v2/application/event_test.go:
```diff
@@ -0,0 +1,86 @@
+		},
+	})
+	_, err := AddEvent(evt, context.Background(), dic)
+	if err != nil {
```
Updated.

lindseysimple commented on internal/core/data/v2/application/event_test.go:
```diff
@@ -0,0 +1,86 @@
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+	if bson.IsObjectIdHex(newId) {
```
Update.

lindseysimple commented on internal/core/data/v2/application/event_test.go:
```diff
@@ -0,0 +1,86 @@
+		},
+	})
+	_, err := AddEvent(evt, context.Background(), dic)
+	if err != nil {
```
Update.

lindseysimple commented on internal/core/data/v2/application/event_test.go:
```diff
@@ -0,0 +1,86 @@
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+	if bson.IsObjectIdHex(newId) {
```
Actually `bson.IsObjectIdHex(newId) == true` is a failure as this is a no persistence testcase so which doesn't expect a `newId`. This testcase is inherited from v1. 
Please let me know if you have further suggestion regarding this testcase, thanks.

lenny-intel replied:
Ok, thanks for the clarification. Looks good with recent updates.

lindseysimple commented on internal/core/data/v2/application/mocks/device.go:
```diff
@@ -0,0 +1,76 @@
+
+// NewMockDeviceClient creates a new mock DeviceClient which has some general mocking behavior defined.
+// will modify this function once core-metadata V2 Device API is available
+func NewMockDeviceClient() *mocks.DeviceClient {
```
File renamed, thanks.

cloudxxx8 commented on internal/core/data/v2/application/device_test.go:
```diff
@@ -0,0 +1,69 @@
+	testEvent.Readings = buildReadings()
+}
+
+func buildReadings() []contract.Reading {
```
it's a function to make dummy readings for unit test only.
It might be better in event_test.go, or it would be included in the compiled binary.

cloudxxx8 commented on internal/core/data/v2/application/device.go:
```diff
@@ -0,0 +1,29 @@
+	"github.com/edgexfoundry/go-mod-bootstrap/di"
+)
+
+// This function will be updated when CheckDevice in v2 core-metadata is available
```
CheckDevice is a function provided by core-metadata client to check whether this device name exists in core-metadata.  It is currently calling v1 API, so we still need to change it once the V2 core-metadata API is available

lenny-intel commented on internal/core/data/v2/application/test_utils.go:
```diff
@@ -0,0 +1,69 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package application
```
As @cloudxxx8 pointed out this file needs to end in `_test.go`. How about `testUtiles_test.go`?

lenny-intel commented on internal/pkg/v2/infrastructure/redis/reading.go:
```diff
@@ -0,0 +1,72 @@
+	if err != nil {
+		return id, err
+	}
+	// use the SET command to save reading as blob
```
V1 has following, which I assume was for adding a reading to existing Event, which we don't support, correct?
```
	if tx {
		_ = conn.Send("MULTI")
	}
....
	if tx {
		_, err = conn.Do("EXEC")
	}
```

lindseysimple replied:
Yes, in V1 the `tx = true`case is invoked while calling the `AddReading` function(in `internal/pkg/db/redis/data.go`) used by the POST Reading API. As we don't have the POST Reading API in V2, we would not need the `tx=true` case, as the `addReading` function is always called in `addEvent`.

lenny-intel commented on internal/pkg/v2/infrastructure/redis/reading.go:
```diff
@@ -0,0 +1,72 @@
+		// Clear the binary data since we do not want to persist binary data to save on memory.
+		newReading.BinaryValue = emptyBinaryValue
+
+		baseReading = &newReading.BaseReading
+		checkReadingValue(baseReading)
+		m, err = json.Marshal(newReading)
```
These are duplicated and can be moved below the switch.  Leaving `case model.SimpleReading:` empty.

lindseysimple replied:
It won't work if the above code is moved out of the switch block since the `newReading` value should be passed as a struct.

andresrinivasan replied:
Ah, I see it now; we need those lines in the switch as that is the small block where we know which concrete type the interface is carrying. Which is the point @lindseysimple was trying to convey earlier. Apologies for my being slow on the uptake.

I did notice the update to `model.Reading` and it got me thinking we can take advantage of closure to perhaps simplify the overall code. I [created a gist](https://gist.github.com/andresrinivasan/0d7429d685008d2f6ed9901aab3183da#file-reading-go) after playing with the code. Outside of any errors I may have created, we are straying into philosophy so please consider this as you would any form of artistic expression.

lenny-intel replied:
Ok, i see now that `newReading` is only in scope of switch. THX!

lindseysimple commented on internal/pkg/v2/infrastructure/redis/event.go:
```diff
@@ -0,0 +1,73 @@
+		e.Id = uuid.New().String()
+	}
+
+	event := model.Event{
```
Use another event variable here as the `Readings` and `isValidated` fields of `model.Event` are not saved in the event blob as in V1(in `marshalEvent` function of `internal/pkg/db/redis/event.go`).

lenny-intel replied:
Got it! Thanks. BTW, `isValidated` is no longer needed on the v2 models since we are doing all the validation on the V2 DTOs.

lindseysimple commented on internal/pkg/v2/infrastructure/redis/event.go:
```diff
@@ -0,0 +1,73 @@
+	rids := make([]interface{}, len(e.Readings)*2+1)
+	rids[0] = EventsCollection + ":readings:" + e.Id
+	for i, r := range e.Readings {
+		readingId, err := addReading(conn, r)
```
The `deviceName` of readings are assigned in the `ToReadingModel` function in https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/v2/dtos/reading.go#L51-L57.

As for the `Created` time of reading which is a timestamp saved in DB is defined in the `checkReadingValue` function
in `internal/pkg/v2/infrastructure/redis/reading.go`
```
func checkReadingValue(b *model.BaseReading) (err error) {
	if b.Created == 0 {
		b.Created = common.MakeTimestamp()
	}
```
https://github.com/edgexfoundry/edgex-go/pull/2610/files#diff-bec3ad46361da938c3c8bb25f916e4d6R58-R61

lenny-intel replied:
I am good on handling of `Created` THX!
Since we don't have any validation on the on the Reading DTO, we are not guaranteed that it is set. 
https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/v2/dtos/reading.go#L25
I am ok with addressing this with more robust validation on the AddEventRequest & Reading DTOs 
If you agree, please submit issue on go-core-contracts for `robust validation on the AddEventRequest & Reading DTOs`. THX.

lindseysimple replied:
Thanks for the suggestion. PRs https://github.com/edgexfoundry/edgex-go/pull/2653 and https://github.com/edgexfoundry/go-mod-core-contracts/pull/264 are opened for the suggested validation enhancements.

lindseysimple commented:
Create PR https://github.com/edgexfoundry/go-mod-core-contracts/pull/262 for `Reading` interface to access `BaseReading` fields. This PR be held until the go-mod-core-contract PR is merged.

lindseysimple commented on internal/pkg/v2/infrastructure/redis/event.go:
```diff
@@ -0,0 +1,73 @@
+			return model.Event{}, err
+		}
+		// set the sorted set score to the index of the reading
+		rids[i*2+1] = i
```
Modify `addReading` function to return the whole reading object.

lenny-intel replied:
Got it! THX!

cloudxxx8 commented:
recheck

lenny-intel commented on internal/core/data/v2/io/event.go:
```diff
@@ -0,0 +1,44 @@
+	c := context.WithValue(*ctx, clients.ContentType, clients.ContentTypeJSON)
+	*ctx = c
+	var addEvents []dto.AddEventRequest
+	err = json.NewDecoder(reader).Decode(&addEvents)
```
Should be using `(a *AddEventRequest) UnmarshalJSON` so Validation is called? or call Validate on each AddEventRequest.

lindseysimple replied:
The `Decode` function would invoke the `unmarshal` function, so eventually the `(a *AddEventRequest) UnmarshalJSON` will be called by the `Decode` function, in which `(a AddEventRequest) Validate()` is invoked in `(a *AddEventRequest) UnmarshalJSON`. 
Hopefully this makes sense, thanks.

lenny-intel replied:
Yes, later I was thinking that might be the case due to the DTO implements the `Unmarshaler` interface.  THX!

lindseysimple commented:
Update the code including:
1. Use the latest `go-mod-core-contracts` version(v0.1.69) to apply the Event and Reading DTO validation.
2. Refactor to define all the Redis commands in the `internal/pkg/v2/infrastructure/redis/const.go` file .

lenny-intel commented on internal/pkg/v2/infrastructure/redis/reading.go:
```diff
@@ -0,0 +1,77 @@
+func addReading(conn redis.Conn, r model.Reading) (reading model.Reading, err error) {
+	var m []byte
+	var baseReading *model.BaseReading
+	switch newReading := r.(type) {
```
Can now use r.ValueType to determine if Binary or Simple, rather than reflection. Always best to avoid reflection if possible.

lindseysimple replied:
@lenny-intel Lenny, If changing to not use type conversion, the code will look like below
```
	if r.GetBaseReading().ValueType == dtos.ValueTypeBinary {
		// Handle BinaryReading value
		// Clear the binary data since we do not want to persist binary data to save on memory.
		newReading := r.(model.BinaryReading)
		newReading.BinaryValue = emptyBinaryValue

		baseReading = &newReading.BaseReading
		if err = checkReadingValue(baseReading); err != nil {
			return nil, err
		}
		m, err = json.Marshal(newReading)
		reading = newReading
	} else {
		// Handle SimpleReading value
		newReading := r.(model.SimpleReading)
		baseReading = &newReading.BaseReading
		if err = checkReadingValue(baseReading); err != nil {
			return nil, err
		}
		m, err = json.Marshal(newReading)
		reading = newReading
	}
```
Will still need to do the type conversion inside the if/else condition and which might cause panic when the `ValueType` is invalid.  Seems that use the original code might be a safer way. Any other suggestion on this? Thanks.

lenny-intel replied:
Somewhere before we get here, we should be validating that `ValueType` is set appropriately and not make it here with an invalid type. I think we should be validating value of `ValueType` here: https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/v2/dtos/reading.go#L64

lenny-intel replied:
Also, I get your point on the type conversion... ;-) Going to have it ether way, so I am good with the switch/case approach.

lenny-intel commented on internal/core/data/v2/controller/http/event.go:
```diff
@@ -0,0 +1,85 @@
+
+	w.Header().Set(clients.ContentType, clients.ContentTypeJSON)
+	w.Header().Set(clients.CorrelationHeader, correlation.FromContext(ctx))
+	w.WriteHeader(http.StatusMultiStatus)
```
FYI, in the App Services I had issues with this call being after the ones above. Please verify you are getting the correct status code returned to caller. I had to move it above the other Header.Sets(). 

Also you should add unit tests similar to these in App Service V2 controller. i.e. there is a test framework in Go for mocking and testing HTTP endpoints.
https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/internal/v2/controller/http/controller_test.go

lindseysimple replied:
Add the controller test case in `event_test.go` as suggested, thanks.

lindseysimple commented:
Added test cases in `internal/core/data/v2/application/device_test.go`  and `internal/core/data/v2/controller/event_test.go`.

lenny-intel commented on internal/core/data/v2/controller/http/event_test.go:
```diff
@@ -0,0 +1,187 @@
+		{"Invalid - No SimpleReading Value", []requests.AddEventRequest{noSimpleValue}, true, http.StatusBadRequest},
+		{"Invalid - No SimpleReading FloatEncoding", []requests.AddEventRequest{noSimpleFloatEnconding}, true, http.StatusBadRequest},
+		{"Invalid - No BinaryReading BinaryValue", []requests.AddEventRequest{noBinaryValue}, true, http.StatusBadRequest},
+		{"Invalid - No BinaryReading MediaType", []requests.AddEventRequest{noBinaryMediaType}, true, http.StatusBadRequest},
```
Add a test case for invalid `valueType` , passing something like `"ValueTypeBad"` as the `valueType`

lindseysimple replied:
Updated to use the latest `go-mod-core-contracts` version(v0.1.72) and added the invalid `valueType` testcase as suggested.
